src/NHibernate/Linq/GroupJoin/NonAggregatingGroupJoinRewriter.cs

```diff
@@ -92,19 +93,30 @@ private void ReWrite()
 			}
 		}
 
-		private void ProcessFlattenedJoin(GroupJoinClause nonAggregatingJoin)
+		private void ProcessFlattenedJoin(GroupJoinClause nonAggregatingJoin, QuerySourceUsageLocator locator)
 		{
+			var nhJoin = locator.LeftJoin
+				? new NhOuterJoinClause(nonAggregatingJoin.JoinClause)
+				: (IQuerySource) nonAggregatingJoin.JoinClause;
+
 			// Need to:
 			// 1. Remove the group join and replace it with a join
 			// 2. Remove the corresponding "from" clause (the thing that was doing the flattening)
-			// 3. Rewrite the selector to reference the "join" rather than the "from" clause
-			SwapClause(nonAggregatingJoin, nonAggregatingJoin.JoinClause);
+			// 3. Rewrite the query model to reference the "join" rather than the "from" clause
+			SwapClause(nonAggregatingJoin, (IBodyClause) nhJoin);
```

Comment:
A bit unrelated.. But do you know any reasons why `SwapClause` method is implemented as `RemoveAt` and `Insert` instead of simply `bodyClauses[i] = newClause`

Reply:
The only difference is in how the collection changed event is triggered, when using an indexer to set a value the event handler will be invoked once by using [NotifyCollectionChangedAction.Replace](https://referencesource.microsoft.com/#system/compmod/system/collections/objectmodel/observablecollection.cs,231),  and with `RemoveAt/Insert` it will be invoked two times. The same is done in `LeftJoinRewriter`:
https://github.com/nhibernate/nhibernate-core/blob/ac39173567b31bcfad475ca32687c9faf0d37f87/src/NHibernate/Linq/Visitors/LeftJoinRewriter.cs#L46-L47
and in `AggregatingGroupByRewriter`:
https://github.com/nhibernate/nhibernate-core/blob/4d47799230686aef2e3377bce42c8b35f7bd3674/src/NHibernate/Linq/GroupBy/AggregatingGroupByRewriter.cs#L86-L87

so maybe we should address this in a seprate PR if needed.

Reply:
Ok.. Yeah I figured this part. I thought you know if such difference is important somehow... Yeah can be fixed in other PR later..

Reply:
>  I thought you know if such difference is important somehow.

To my knowledge it is not important and it should work the same.